libraries/chain/database.cpp

```diff
@@ -5410,31 +5421,66 @@ void database::retally_witness_vote_counts( bool force )
-   // Note that no decimal places argument is required from SMT creator at this stage.
-   // This is because asset_symbol_type's to_string method omits the precision when serializing.
-   // For appropriate use of this method see e.g. smt_database_fixture::create_smt
+   const auto& idx = get_index< nai_pool_index >().indices().get< by_symbol >();
 
-   uint8_t decimal_places = 0;
+   vector< asset_symbol_type > nai_pool;
+   for ( auto it = idx.begin(); it != idx.end(); ++it )
+      nai_pool.push_back( it->asset_symbol );
 
-   FC_ASSERT( _next_available_nai >= SMT_MIN_NON_RESERVED_NAI );
-   FC_ASSERT( _next_available_nai <= SMT_MAX_NAI, "Out of available NAI numbers." );
-   // Assume that _next_available_nai value shows the liquid version of NAI.
-   FC_ASSERT( (_next_available_nai & 0x1) == 0, "Can't start with vesting version of NAI." );
-   uint32_t new_nai = _next_available_nai;
-   // Skip vesting version of produced NAI - it differs only by least significant bit set.
-   _next_available_nai += 2;
+   return nai_pool;
+}
 
-   uint32_t new_asset_num = (new_nai << 5) | 0x10 | decimal_places;
-   asset_symbol_type new_symbol = asset_symbol_type::from_asset_num( new_asset_num );
-   new_symbol.validate();
-   FC_ASSERT( new_symbol.space() == asset_symbol_type::smt_nai_space );
+void database::replenish_nai_pool()
+{
+   try
+   {
+      nai_generator n;
+      while ( count< nai_pool_object >() < SMT_MAX_NAI_POOL_COUNT )
+      {
+         asset_symbol_type next_sym;
+         auto block_id = head_block_id();
+         auto seed = block_id._hash[0];
+         uint32_t collisions = 0;
+         do
+         {
+            if ( collisions++ >= SMT_MAX_NAI_GENERATION_TRIES )
+            {
+               wlog( "Encountered ${collisions} collisions while attempting to generate NAI",
+                    ("collisions", collisions) );
+               return;
+            }
+            next_sym = n( seed++ );
+         }
+         while ( asset_symbol_exists_in_nai_pool( next_sym ) || asset_symbol_is_an_smt_token( next_sym ) );
+
+         create< nai_pool_object >( [&]( nai_pool_object& npo ) {
+            npo.asset_symbol = next_sym;
+         });
 
-   return vector< asset_symbol_type >( 1, new_symbol );
+         FC_ASSERT( asset_symbol_exists_in_nai_pool( next_sym ), "Asset symbol should exist in NAI pool." );
+      }
+   }
+   FC_CAPTURE_AND_RETHROW()
 }
+
+bool database::asset_symbol_exists_in_nai_pool( asset_symbol_type a )
+{
+   const auto& idx = get_index< nai_pool_index >().indices().get< by_symbol >();
+   auto stripped_symbol_num = a.get_stripped_precision_smt_num();
+   auto it = idx.lower_bound( asset_symbol_type::from_asset_num( stripped_symbol_num ) );
+   return it != idx.end() && it->asset_symbol.get_stripped_precision_smt_num() == stripped_symbol_num;
+}
+
+bool database::asset_symbol_is_an_smt_token( asset_symbol_type a )
```

Comment:
Is there any way we can rewrite this to not live in the database? Considering how short the method is, it seems like this could just be inlined wherever needed...